### ISSUE TITLE
PURE Database and Data loader Issue

### ISSUE THREAD

**Issue by Taimoor-R** [ubicomplab/rPPG-Toolbox#109]
Hi, i am wondering how you set up the PURE database as it has a json file containing frame information and pgg data but the ppg timestamps dont correspond to image timestamps. So how do you get the coresponding ppg signal to the image frame.

**Comment by yahskapar** [comment]
The label (PPG signal ground truth) is re-sampled to match the video using the sample() function found in this toolbox (utils/utils.py). That's done shortly after the frames and signal are read in [here](https://github.com/ubicomplab/rPPG-Toolbox/blob/main/dataset/data_loader/PURELoader.py#L114).

**Comment by Taimoor-R** [comment]
Thank you for the prompt response I have a follow up question: 
The dataloader tensor does it look like this: 
Frame(x) shape :: torch.Size([batch, frames , channel, height, width]) Label(y) shape :: torch.Size([batch])

**Comment by Taimoor-R** [comment]
Basically i want to use the PURE dataset with a model i am buliding myself but i cant seem to understand how to format the data so for example:

I want to load 16 32 frame clips along 32 corresponding gt labels of each clip. Then i want to predict 32 values for the 32 frame clip I pass to my model. So esentially i just want to understand how to create a dataloader for that purpose.
Basically i want to use the PURE dataset with a model i am buliding myself but i cant seem to understand how to format the data so for example:

I want to load 16 32 frame clips along 32 corresponding gt labels of each clip. Then i want to predict 32 values for the 32 frame clip I pass to my model. So esentially i just want to understand how to create a dataloader for that purpose.
Basically, I want to use the PURE dataset with a model I am building myself, but I can't seem to understand how to format the data so for example:

I want to load 8/16(batch size i think) 32 frame clips along 32 corresponding gt labels of each clip. Then I want to predict 32 values for the 32 frame clip I pass to my model. So essentially I just want to understand how to create a data loader for that purpose.

**Comment by yahskapar** [comment]
I see - the code I pointed to in my previous reply should be helpful to that end, but note again that it's specific to the neural method TS-CAN. It may be helpful to read a bit of the [MTTS-CAN paper](https://papers.nips.cc/paper/2020/file/e1228be46de6a0234ac22ded31417bc7-Paper.pdf) itself to get a better idea of why the code to process the inputs for the model is the way it is.

Also, if you aren't doing this already, you may find it helpful to directly use the rPPG-Toolbox in your own fork and just create new files related to your neural method (e.g., a trainer file for your 3D CNN model) based on the existing neural method files. That way you can leverage preprocessing code, and possibly other code, from the rPPG-Toolbox directly and prevent bugs from popping up if you happen to be using segments of code in your own, completely separate project code.

**Comment by Taimoor-R** [comment]
Thank you @yahskapar I will surely fork the repo and then use it. I just had a request if you could look at the following loss graphs(training and Validation on PURE): 
![Screenshot 2023-01-24 at 23 57 03](https://user-images.githubusercontent.com/70077383/214447695-8e9f4b9b-834d-4eb9-a800-a99341a781ca.png)
I just wanted to ask do the results look ok or am I doing something wrong in understanding the dataloaders.

**Comment by yahskapar** [comment]
Assuming your x-axis is the number of epochs, those curves look reasonable to me.

**Comment by Taimoor-R** [comment]
Is it normal for the loss(MSE) not to drop for first 20 epochs and then start to drop?

**Comment by girishvn** [comment]
Hi @Taimoor-R, can you specify what model architecture you are using? Generally for models such as TSCAN we see validation losses (MSE) of at most ~0.8 by training Epoch 5. The fact that your validation loss remains over 1 for so many epochs seems strange.

**Comment by girishvn** [comment]
I would suggest first running PURE_PURE_UBFC_TSCAN to ensure that the training/validation/test pipeline are working correctly. You should get fairly good FFT-based HR results within the first 10 epochs

**Comment by Taimoor-R** [comment]
Hi @girishvn I am implementing my own architecture(The loss-epoch graphs above) as well as a video transformer based architecture(no change in loss even after 30 epochs), I think I am probably doing something wrong but I cant really pinpoint where I am going wrong. Furthermore the graph above is just a Training and Validation losses on PURE dataset(preprocessed using code from this repo- chunk-size = 32, batch = 16, H = 128, W = 128). If you could guide me a bit I would be grateful

**Comment by girishvn** [comment]
I would suggest, that you first attempt to train a toolbox supported model w/ PURE, to ensure that you are using the toolbox pipeline correctly. This will help isolate if this is an network architecture problem, or a preprocessing/toolbox problem.

If you see reasonable train and validation loss curves w/ a toolbox supported model and PURE, then the problem is likely the network architecture or some modeling hyper param (as opposed to preprocessing / your trainer code)

**Comment by Taimoor-R** [comment]
Thank you I will try to do that and check

**Comment by yahskapar** [comment]
To add onto what @girishvn mentioned, be sure to make sure you have the latest changes in this repo before using it. Quite a few changes were made in the past few weeks, some of which may affect whatever code you have in place as a part of some larger pipeline (that may or may not currently include the toolbox based on what you've said).

**Comment by Taimoor-R** [comment]
Hi again @yahskapar @girishvn, I have run physnet from the lastest release of the rPPG toolbox, can you please take a look at the graphs as they look very odd espically the validation loss curve.
![Screenshot 2023-01-30 at 20 15 57](https://user-images.githubusercontent.com/70077383/215585876-a36896c4-6fb7-4c40-919b-117c9e9357b7.png)

**Comment by xliucs** [comment]
@Taimoor-R  Did you use the training schedule we provided? If so, use the last epoch should be fine. In the config file, it should be trained with 30 epochs.

**Comment by Taimoor-R** [comment]
@xliucs I did use the training schedule but changed the number of epochs to see if the training curve would be comparable to the fitness curve. As I am implementing my own method I wanted to understand what is a reasonable loss for training and validation is

**Comment by Taimoor-R** [comment]
> @Taimoor-R Did you use the training schedule we provided? If so, use the last epoch should be fine. In the config file, it should be trained with 30 epochs.

Additionally I want to understand what you mean by use the last epoch as in the 30th or use last epoch as in disregard validation. As I am not testing this model yet just running training and validation.

**Comment by xliucs** [comment]
You can use the best mdoel from validation loss by setting `USE_LAST_EPOCH` to `False`.  50 epochs might be too many.

**Comment by yahskapar** [comment]
> @xliucs I did use the training schedule but changed the number of epochs to see if the training curve would be comparable to the fitness curve. As I am implementing my own method I wanted to understand what is a reasonable loss for training and validation is

What do you mean by "fitness curve"? 

Also, just to be sure, are you actually running PhysNet **within** the latest rPPG-Toolbox (and through the corresponding YAML config file), or did you re-use the implementation of that network and its trainer module elsewhere? I want to make sure you are actually using the toolbox as is, including the OneCycleLR scheduler used in the toolbox's trainer modules. There are numerous factors that can affect the loss and validation curves (e.g., the dataset used, learning rate specified as the maximum in the YAML config, and number of epochs, among other things). Can you specify what dataset, or datasets, you are using for training and validation, and exactly what data (e.g., subject-wise) composes the training and validation splits? Also, the learning rate, number of epochs,  batch size, and the number of GPUs - feel free to just paste the YAML config you used here. I can try to reproduce what you are doing on my end sometime later this week.

For what it's worth, those curves you provided might be reasonable (within the context of the number of epochs you specified, of course). The validation loss curve may look a little janky, but it makes sense given how OneCycleLR works and appears to stabilize in the last few epochs, as one would expect since the learning rate is so small at that point. You can read more about OneCycleLR [here](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.OneCycleLR.html) and [here](https://towardsdatascience.com/finding-good-learning-rate-and-the-one-cycle-policy-7159fe1db5d6).
